src/gleam/list.gleam

```diff
@@ -248,6 +249,48 @@ pub fn rest(list: List(a)) -> Result(List(a), Nil) {
   }
 }
 
+fn group_updater(
+  f: fn(element) -> key,
+) -> fn(Map(key, List(element)), element) -> Map(key, List(element)) {
+  fn(groups, elem) {
+    case map.get(groups, f(elem)) {
+      Ok(existing) -> map.insert(groups, f(elem), [elem, ..existing])
+      Error(_) -> map.insert(groups, f(elem), [elem])
+    }
+  }
+}
+
+/// Takes a list and groups the values by a key
+/// which is built from a key function.
+///
+/// ## Examples
+///
+/// ```gleam
+/// > [Ok(3), Error("Wrong"), Ok(200), Ok(73)]
+///   |> group(with: fn(i) {
+///     case i {
+///       Ok(_) -> "Successful"
+///       Error(_) -> "Failed"
+///     }
+///   })
+///   |> map.to_list
+///
+/// [
+///   #("Failed", [Error("Wrong")]),
+///   #("Successful", [Ok(3), Ok(200), Ok(73)])
+/// ]
+///
+/// > group(from: [1,2,3,4,5], with: fn(i) {fn(i) { i - i / 3 * 3 }})
+/// |> map.to_list
+/// [#(0, [3]), #(1, [1, 4]), #(2, [2, 5])]
+/// ```
+///
+pub fn group(list: List(v), by key: fn(v) -> k) -> Map(k, List(v)) {
+  list
+  |> fold(map.new(), group_updater(key))
```

Comment:
Could we have a more descriptive name for this function please?

Reply:
I actually took this name from the `iterator.group` implementation. But I changed it.